contracts/smart-account/modules/BatchedSessionRouterModule.sol

```diff
@@ -0,0 +1,190 @@
+                b) verifies the session key permissions via Session Validation Modules
+ *         - Should be used with carefully verified and audited Session Validation Modules only
+ *         - Compatible with Biconomy Modular Interface v 0.1
+ * @author Fil Makarov - <filipp.makarov@biconomy.io>
+ */
+
+contract BatchedSessionRouter is BaseAuthorizationModule {
+    bytes4 public constant EXECUTE_BATCH_SELECTOR = 0x47e1da2a;
+    bytes4 public constant EXECUTE_BATCH_OPTIMIZED_SELECTOR = 0x00004680;
+
+    /**
+     * @dev validates userOperation. Expects it to be a executeBatch or executeBatch_y6U call
+     * If something goes wrong, reverts
+     * @param userOp User Operation to be validated.
+     * @param userOpHash Hash of the User Operation to be validated.
+     * @return SIG_VALIDATION_FAILED or packed validation result.
+     */
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external virtual returns (uint256) {
+        // check this is a proper method call
+        require(
+            bytes4(userOp.callData[0:4]) == EXECUTE_BATCH_OPTIMIZED_SELECTOR ||
+                bytes4(userOp.callData[0:4]) == EXECUTE_BATCH_SELECTOR,
+            "SR Invalid Selector"
+        );
+
+        (bytes memory moduleSignature, ) = abi.decode(
+            userOp.signature,
+            (bytes, address)
+        );
+
+        /*
+        struct SessionCallData {
+            uint48 validUntil;
+            uint48 validAfter;
+            address sessionValidationModule;
+            bytes sessionKeyData;
+            bytes32[] merkleProof;
+            bytes callSpecificData;
+        }
+        */
+
+        // parse the signature to get the array of required parameters
+        (
+            address sessionKeyManager,
+            uint48[] memory validUntil,
+            uint48[] memory validAfter,
+            address[] memory sessionValidationModule,
+            bytes[] memory sessionKeyData,
+            bytes32[][] memory merkleProof,
+            bytes[] memory callSpecificData,
+            bytes memory sessionKeySignature
+        ) = abi.decode(
+                moduleSignature,
+                (
+                    address,
+                    uint48[],
+                    uint48[],
+                    address[],
+                    bytes[],
+                    bytes32[][],
+                    bytes[],
+                    bytes
+                )
+            );
+
+        // can I use struct instead so don't have to check lengths
+
+        // check lengths of arrays
+        require(
+            validUntil.length == validAfter.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == sessionValidationModule.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == sessionKeyData.length,
+            "SR Invalid data provided"
+        );
+        require(
```

Comment:
Do we not need a length check for `callSpecificData`?

Reply:
also a check against the length of the batch

Reply:
this block was rebuilt to use struct instead and no checks required